.github/workflows/test.yml

```diff
@@ -11,44 +11,48 @@ on:
 
 jobs:
   golint:
-    name: vet fmt and lint
-    runs-on: ubuntu-20.04
+    name: Lint
+    runs-on: ubuntu-latest
     container: golang:1.13.7
     steps:
     - uses: actions/checkout@v2.3.3
     - name: Install golint
       run: go get -u golang.org/x/lint/golint
     - run: make lint
       shell: bash
+
   gotest:
-    name: go test
-    runs-on: ubuntu-20.04
+    name: Test
+    runs-on: ubuntu-latest
     container: golang:1.13.7
     steps:
     - uses: actions/checkout@v2.3.3
     - run: make test
+
   docker_build:
-    name: docker_build
-    runs-on: ubuntu-20.04
-    container:
-      image: docker:20.10.5
-      options: -t
-    # Setting up dind service container
-    services:
-      docker:
-        image: docker:20.10.5-dind
-        env:
-          DOCKER_DRIVER: overlay
-          DOCKER_HOST: tcp://localhost:2375
+    name: Docker
+    runs-on: ubuntu-latest
     steps:
-    - name: Install Tools
-      run: apk add --update make git jq rsync curl
-    - uses: actions/checkout@v2.3.3
-    - name: Set up QEMU
-      uses: docker/setup-qemu-action@v1
-    - name: Set up Docker Buildx
-      uses: docker/setup-buildx-action@v1
-      with:
-        driver-opts: image=moby/buildkit:master
-    - name: Build images
-      run: make build-docker-image
+      # Pull down source code
+      - name: Checkout
+        uses: actions/checkout@v2.3.3
+
+      # Set up docker-buildx for multi-arch images
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v1
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+
+      # Install golang 1.13
+      - name: Install Go
+        uses: actions/setup-go@v2
+        with:
+          go-version: 1.13.7
+
+      # Compile multi-arch binaries
+      - name: Compile binaries
+        run: make build
```

Comment:
I wonder how important this is. Is it just confusing that we publish windows and mac images. Just because we can doesn't mean we should.

![Yeah, yeah, but your scientists were so preoccupied with whether or not  they could that they didn't stop to… | Jurassic park quotes, Park quotes,  Best movie quotes](https://user-images.githubusercontent.com/1774239/116562392-050e1a00-a8fb-11eb-85aa-d5824966c7b3.jpeg)

I wonder if we can skip this step and instead copy the binary from the built image, docker create, docker cp etc.

Reply:
Do we need to do this in test anyway?

Reply:
So this workflow doesn't actually push any images, it just ensures that the docker image builds. The reason for compiling the binaries here is that all the docker image does now is copy the built binary into itself. So we need to build the binaries on the host VM first, so that `docker build` will pick them up.

It doesn't necessarily need to exist, however it might be nice to catch things that break the docker build in PRs rather than once merged to master.

Reply:
> is that all the docker image does now is copy the built binary into itself

How does that work for multiple architectures. I can't see where that happens.

Reply:
So for each architecture, the image is built:

```Dockerfile
# TARGETPLATFORM comes from the buildx context and it will be something like `linux/arm64/v8` or `linux/amd64`.
# Ref: https://docs.docker.com/buildx/working-with-buildx/
ARG TARGETPLATFORM
```

The `TARGETPLATFORM` is automatically set by the `buildx` tool and we use it to build the path to the correct binary, which is copied in.

```Dockerfile
COPY ./builds/${TARGETPLATFORM}/preflight /bin/preflight
```

When binaries are compiled, they're placed into a directory tree like `OS/ARCH/preflight`